mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -178,67 +206,95 @@ private Connection getConnectionInternal(int connectionType) {
-                Message result = mMainHandler.obtainMessage(MSG_UPDATE_DONE);
+            case MSG_TRIGGER_UPDATE: { // update thread
                 Connection local, remote;
                 synchronized (this) {
                     local = mLocalConnection;
                     remote = mRemoteConnection;
                 }
+                ConnectionUpdateResult result = new ConnectionUpdateResult();
                 try {
-                    result.obj = determineAvailableConnection(ctx, local, remote);
+                    result.available = determineAvailableConnection(ctx, local, remote);
                 } catch (ConnectionException e) {
-                    result.obj = e;
+                    result.availableFailureReason = e;
                 }
-                mMainHandler.sendMessage(result);
-                return true;
-            case MSG_UPDATE_DONE: // main thread
-                if (msg.obj instanceof ConnectionException) {
-                    updateAvailableConnection(null, (ConnectionException) msg.obj);
+                if (remote != null) {
+                    mUpdateHandler.obtainMessage(MSG_UPDATE_CLOUD_CONNECTION, result).sendToTarget();
                 } else {
-                    // Check whether the passed connection matches a known one. If not, the
-                    // connections were updated while the thread was processing and we'll get
-                    // a new callback.
-                    if (msg.obj == mLocalConnection || msg.obj == mRemoteConnection) {
-                        updateAvailableConnection((Connection) msg.obj, null);
-                    }
+                    mMainHandler.obtainMessage(MSG_UPDATE_DONE, result).sendToTarget();
+                }
+                return true;
+            }
+            case MSG_UPDATE_CLOUD_CONNECTION: { // update thread
+                ConnectionUpdateResult result = (ConnectionUpdateResult) msg.obj;
+                final AbstractConnection remote;
+                synchronized (this) {
+                    remote = mRemoteConnection;
                 }
+                if (remote != null) {
+                    final MySyncHttpClient client = remote.getSyncHttpClient();
+                    client.get(CloudConnection.SETTINGS_ROUTE, new MyHttpClient.TextResponseHandler() {
+                        @Override
+                        public void onFailure(Call call, int statusCode, Headers headers, String responseBody, Throwable error) {
+                            Log.e(TAG, "Error loading notification settings: " + error.getMessage());
+                        }
+
+                        @Override
+                        public void onSuccess(Call call, int statusCode, Headers headers, String responseBody) {
+                            try {
+                                JSONObject json = new JSONObject(responseBody);
+                                String senderId = json.getJSONObject("gcm").getString("senderId");
```

Comment:
As already stated somewhere else: I don't think, that this belongs into the Connection object. And if it does, I would argue, that it should be part of the Connection interface, so that no cast is needed.

Or, when I think about it: If a connection to specific server has special features/parameters, it should probably be it's own type (class or interface). However, it feeld strange to shift the need to cast the connection object to the caller of the COnnectionFactory, though :(

Reply:
> And if it does, I would argue, that it should be part of the Connection interface, so that no cast is needed.

I really have to disagree here: why move a property *specific to one particular type of connection* to the base class? This really only leads to method bloat.
If you're unhappy with the cast, this can be solved by splitting `ConnectionFactory.getConnection()` into 3 separate methods to fetch local/remote/cloud connection. Personally, I think the cast is fine.